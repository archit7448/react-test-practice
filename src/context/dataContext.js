import { createContext, useContext, useReducer } from "react";
import { reducer } from "../reducer/reducer";

const intialValue = {
  products: [
    { id: "123", productName: "Milk", price: "50" },
    { id: "124", productName: "Butter", price: "70" },
    { id: "125", productName: "Eggs", price: "60" },
    { id: "126", productName: "water", price: "20" },
  ],
  cart: [],
};

const DataContext = createContext(null);

const DataProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, intialValue);
  return (
    <DataContext.Provider value={{ state, dispatch }}>
      {children}
    </DataContext.Provider>
  );
};

const useData = () => useContext(DataContext);

export { DataProvider, useData, DataContext };
