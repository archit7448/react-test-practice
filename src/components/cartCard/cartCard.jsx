import { useData } from "../../context/dataContext";

export function CartCard({ onClick }) {
  const { state, dispatch } = useData();
  const { cart } = state;
  return (
    <div>
      {state.products.map(({ productName, price, id }) => {
        return (
          <div key={id}>
            <h1 data-testid="product-name">{productName}</h1>
            <h2>{price}</h2>
            <button
              onClick={() => {
                dispatch({ type: "ADD_TO_CART", payload: { id: id } });
                onClick();
              }}
            >
              Add to Cart
            </button>
          </div>
        );
      })}
    </div>
  );
}
