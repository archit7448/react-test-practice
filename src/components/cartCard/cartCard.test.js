import { cleanup, render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { DataProvider } from "../../context/dataContext";
import { CartCard } from "./cartCard";

afterEach(cleanup);
describe("Card Cart", () => {
  test("render product", () => {
    const { getAllByTestId } = render(<CartCard />, {
      wrapper: DataProvider,
    });

    const productNames = getAllByTestId(/product-name/i).map(
      (div) => div.textContent
    );

    expect(productNames).toMatchSnapshot();
  });

  test("test for button click", () => {
    const cartHandler = jest.fn();
    const { getAllByText } = render(<CartCard onClick={cartHandler} />, {
      wrapper: DataProvider,
    });

    act(() => getAllByText(/add to cart/i).forEach((value) => value.click()));
    expect(cartHandler).toBeCalledTimes(getAllByText(/add to cart/i).length);
  });
});
