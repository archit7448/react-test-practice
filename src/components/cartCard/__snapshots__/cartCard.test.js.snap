// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Cart render product 1`] = `
Array [
  "Milk",
  "Butter",
  "Eggs",
  "water",
]
`;
